src/model_theory/basic.lean

```diff
@@ -313,5 +331,242 @@ lemma comp_assoc (f : M ≃[L] N) (g : N ≃[L] P) (h : P ≃[L] Q) :
 
 end equiv
 
+section closed_under
+
+open set
+
+variables {n : ℕ} (f : L.functions n) (s : set M)
+
+/-- Indicates that a set in a given structure is a closed under a function symbol. -/
+def closed_under : Prop :=
+∀ (x : fin n → M), (∀ i : fin n, x i ∈ s) → fun_map f x ∈ s
+
+variables {f} {s} {t : set M}
+
+lemma closed_under_inter (hs : closed_under f s) (ht : closed_under f t) :
+  closed_under f (s ∩ t) :=
+λ x h, mem_inter (hs x (λ i, mem_of_mem_inter_left (h i)))
+  (ht x (λ i, mem_of_mem_inter_right (h i)))
+
+lemma closed_under_inf (hs : closed_under f s) (ht : closed_under f t) :
+  closed_under f (s ⊓ t) := closed_under_inter hs ht
+
+variables {S : set (set M)}
+
+lemma closed_under_Inf (hS : ∀ s, s ∈ S → closed_under f s) : closed_under f (Inf S) :=
+λ x h s hs, hS s hs x (λ i, h i s hs)
```

Comment:
Dot notation would allow us to write `hs.inter ht` rather than `closed_under_inter hs ht`:
```suggestion
lemma closed_under.inter (hs : closed_under f s) (ht : closed_under f t) :
  closed_under f (s ∩ t) :=
λ x h, mem_inter (hs x (λ i, mem_of_mem_inter_left (h i)))
  (ht x (λ i, mem_of_mem_inter_right (h i)))

lemma closed_under.inf (hs : closed_under f s) (ht : closed_under f t) :
  closed_under f (s ⊓ t) := closed_under_inter hs ht

variables {S : set (set M)}

lemma closed_under.Inf (hS : ∀ s, s ∈ S → closed_under f s) : closed_under f (Inf S) :=
λ x h s hs, hS s hs x (λ i, h i s hs)
```